contracts/core/Stake.sol

```diff
@@ -341,41 +436,107 @@ contract Stake is StakeInterface {
         validatorWeight_ = validatorWeightAtHeight(_height, _validator);
     }
 
+    /**
+     * @notice Returns all registered validator addresses.
+     *
+     * @return An array of validator addresses. Includes evicted validators.
+     */
+    function validatorAddresses() external view returns (address[]) {
+        return validatorAddresses;
+    }
+
     /* Private Functions */
 
+    /**
+     * @notice Calculates the total weight of all active validators at a given
+     *         height.
+     *
+     * @param _height For which height to get the total weight.
+     *
+     * @return totalWeight_ At the given height.
+     */
+    function totalWeightAtHeight_(
+        uint256 _height
+    )
+        private
+        view
+        returns (uint256 totalWeight_)
+    {
+        for (uint256 i = 0; i < validatorAddresses.length; i++) {
+            address validator = validatorAddresses[i];
+            totalWeight_ += validatorWeightAtHeight(_height, validator);
+        }
+    }
+
+    /**
+     * @notice Verifies a new validator to be valid. Includes transfer of the
+     *         deposit amount from the depositor to this contract.
+     *
+     * @param _depositor The address that the deposit is transferred from.
+     * @param _validator The address of the new validator.
+     * @param _amount The initial deposit of the new validator.
+     */
+    function verifyNewValidator(
+        address _depositor,
+        address _validator,
+        uint256 _amount
+    )
+        private
+    {
+        require(
+            _validator != address(0),
+            "The validator address may not be zero."
+        );
+        require(
+            _amount > 0,
+            "The deposit amount must be greater than zero."
+        );
+        require(
+            !validatorExists(_validator),
+            "You must deposit for a validator that is not staked yet."
+        );
+
+        require(
+            stakingToken.transferFrom(_depositor, address(this), _amount),
+            "Could not transfer deposit to the stake contract."
+        );
+    }
+
     /**
      * @notice Creates a new validator and adds it to the list of validators.
      *
-     * @dev You must check that the validator is not yet part of the validators
-     *      before you call this method.
+     * @dev You must verify the validator  before you call this method.
      *
+     * @param _depositor The address that the deposit is transferred from.
      * @param _validator The address of the new validator.
-     * @param _amount The initial stake of the new validator.
+     * @param _amount The initial deposit of the new validator.
+     * @param _startingHeight The height at which the validator will enter the
+     *                        set of validators.
      *
      * @return The index under which the new validator is stored.
      */
     function registerNewValidator(
+        address _depositor,
```

Comment:
Why `_depositor ` address is saved in storage? 
will this be used for during `logout` ?

Reply:
`address` is a value type, not a complex type. That means by default, solidity puts it on the EVM *stack:* [Relevant solidity FAQ](https://solidity.readthedocs.io/en/v0.4.24/frequently-asked-questions.html#what-is-the-memory-keyword-what-does-it-do).
Value types are always copied when they are passed as function arguments (in memory).

Only complex types (arrays and structs) have the optional data location to be specified as `storage` or `memory`. I cannot specify `address storage` or `address memory`, for example.

Furthermore, the default for function arguments would be `memory`, not `storage`.

Regardless, `_depositor` *should* be checked during `logout`. It is stored in the validator:
```solidity
Validator memory validator = Validator(
    _depositor,
    _validator,
    _amount,
    _startingHeight,
    false,
    0
);

validators[_validator] = validator;
```
And later checked to be the correct one when logging out. Something like (wild guess):
```solidity
require(
    msg.sender == validator.depositor,
    "Message sender must be depositor of validator."
);
```